CONTRIBUTING.md

```diff
@@ -2,51 +2,59 @@
 We want to make contributing to this project as easy and transparent as
 possible.
 
-## Pull Requests
-We actively welcome your pull requests.
-
-1. Fork the repo and create your branch from `master`.
-2. If you've added code that should be tested, add tests.
-3. If you've changed APIs, update the documentation.
-4. Ensure the test suite passes.
-5. Make sure your code lints.
-6. If you haven't already, complete the Contributor License Agreement ("CLA").
+## Developer workflow
+- Active development should happen on your fork of the repositories
+- Name your PR in a way that unambiguously identifies the feature or fix
+- Follow the contribution guide to ensure code is conformant with conventions and style
+- Try to make small, logically independent, self-documenting commits (and reflect this in the commit messages by providing brief rationale/change summary). Feel free to amend and force push changes to your PR if necessary.
```

Comment:
I think we shouldn't specifically allow force push/amends on PR. The person who is merging should squash all of the commits later on before merge. This eases the reviewer's pain since they are able to compare what exact difference has been made from last commit based on reviews. (I have felt this pain multiple times when reviewing PRs when people force push).

Reply:
@apsdehal 
> we shouldn't specifically allow force push/amends on P
That contradicts to `Try to make small, logically independent, self-documenting commits (and reflect this in the commit messages by providing brief rationale/change summary)`

Ideally, we would like to have clean break of a larger PR in smaller commits, not history of commit changes on PR. That's hard to achieve without `push -f`.

Reply:
In PR, it is fine. You anyways squash them later on. But for reviewer if you do `push -f` it makes hard to notice exact changes. For e.g. see https://github.com/facebookresearch/habitat-sim/pull/224/commits, it is easy and nice for the reviewer to review each logical chunk and understand what exact difference was there between the two reviews. Otherwise, if I `push -f`, I lose this information.

Reply:
Also, you can be smarter on where you `push -f` and where you don't. That way it is easy.

Reply:
I can remove mentioning about `push -f`, but that's what was the agreement on engineering meeting that we prefer logic breakage of commits based on introduced functionality/changes, rather than historical addressing of reviews. Both approaches have pros and cons, but in the end when accepting a PR that's good to review all changes again, rather than delta from previous review.

Reply:
It is still unclear to me what you mean when you say that what I am saying won't be logical breakage of commits. Let me explain my points again:
1. You can still see the full change in the `Files Changed` tab. So, if one is looking to make a review on full PR they can do it from `Files Changed` tab.
2. As a reviewer, I would want to see what are the changes from the previous time I reviewed  the PR, which is what the latest commit will tell me. 
3. For logic breakage of commit, you would squash all of the commits from the UI into single commit and then merge them to master. So, the feature/functionality wise breakage of commits is still there. Each PR goes into master as single commit. Here is a [pointer](https://medium.com/@vanshbadkul/squashing-commits-in-git-a-step-by-step-guide-884581224adf) on how git squash works.

Reply:
I understand your points and think they are legit, but there are another views as well.

Your proposal:
```
PR Commits:
- Func A
- Func B
- Changes based on review
- Changes based on review
```

We expect:
```
PR Commits:
- Func A
- Func B
```

Some arguments to support last approach:
1. It makes clear what changes are needed for Func A and for Func B in the end of work on PR. And makes it easier to review changes commit by commit.
2. It answers how to revert FuncA and FuncB separately.
3. Comment a code through commit messages in more granular and consistent way.  

Yes, we squash commits in the end, but using github UI you can visit referenced PR and check commit history there.

That's light recommendation as that's very minor improvement for dev efficiency. After your arguments I think that's good to remove if it can raise questions from contributors.

Reply:
What worked for me on magnum very well, is having "style 1" during the time when the PR is being finished (so when I get a new PR push notification, it leads me to the "fixup" commits pushed recently), but then when the PR is ready to be merged, it's rebased and reorganized to "style 2" ... and then put to master *without* squashing, preserving the final clean ordered history *forever*.

However because here the policy is squash-merge, requiring reorganization to "style 2" would be just a total waste of time, since it ends up being a single commit in either case ... and as is usual, various problems get discovered *after* the PR is squash-merged.

Also not everybody is a Git ninja, so demanding one or the other from contributors (as opposed to us the main developers) might be too harsh, maybe? (but also I'm not sure what's the target audience of the `CONTRIBUTING.md` file -- us or 3rd party contributors?) In Magnum itself if the contributor is incapable of that, I'm doing the history cleanup during merge (which is time-consuming, but paid off many times).

> Yes, we squash commits in the end, but using github UI you can visit referenced PR and check commit history there.

Unfortunately that's not really working well for `git blame` (even in the UI) -- all you'd see is the whole file being added in a single commit, and as far as I know, there's no way to do a blame on a PR alone. This goes together with my other concerns about squashing commits (and losing important history that way), but that's for a longer discussion.